chain-impl-mockchain/benches/tally.rs

```diff
@@ -201,27 +258,130 @@ fn tally_benchmark(voters_count: usize, voting_power_per_voter: u64, c: &mut Cri
     ledger.apply_fragment(&fragment, ledger.date()).unwrap();
 }
 
+fn tally_benchmark_flat_distribution(
+    voters_count: usize,
+    voting_power_per_voter: u64,
+    c: &mut Criterion,
+) {
+    let voting_power_distribution = rand::distributions::uniform::Uniform::from(
+        voting_power_per_voter..voting_power_per_voter + 1,
+    );
+    tally_benchmark(1, voters_count, 1.0, 0.5, voting_power_distribution, c);
+}
+
 fn tally_benchmark_128_voters_1000_ada(c: &mut Criterion) {
-    tally_benchmark(128, 1000, c);
+    tally_benchmark_flat_distribution(128, 1000, c);
 }
 
 fn tally_benchmark_200_voters_1000_ada(c: &mut Criterion) {
-    tally_benchmark(200, 1000, c);
+    tally_benchmark_flat_distribution(200, 1000, c);
 }
 
 fn tally_benchmark_200_voters_1_000_000_ada(c: &mut Criterion) {
-    tally_benchmark(200, 1_000_000, c);
+    tally_benchmark_flat_distribution(200, 1_000_000, c);
 }
 
 fn tally_benchmark_1000_voters_1000_ada(c: &mut Criterion) {
-    tally_benchmark(1000, 1000, c);
+    tally_benchmark_flat_distribution(1000, 1000, c);
+}
+
+struct FundDistribution<'a, 'b> {
+    threshold: u64,
+    ranges_bounds: &'a [u64],
+    ranges_weights: &'b [f64],
+}
+
+impl<'a, 'b> Distribution<u64> for FundDistribution<'a, 'b> {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u64 {
+        assert_eq!(self.ranges_weights.len(), self.ranges_bounds.len());
+        let ranges_sampler = WeightedIndex::new(self.ranges_weights).unwrap();
+        let range_no = ranges_sampler.sample(rng);
+        let lower_bound = range_no
+            .checked_sub(1)
+            .map(|i| self.ranges_bounds[i])
+            .unwrap_or(self.threshold);
+        let current_range_sampler = self
+            .ranges_bounds
+            .get(range_no)
+            .copied()
+            .map(|upper_bound| Uniform::from(lower_bound..upper_bound))
```

Comment:
Need a discussion on this. Ideally, we do not need an upper bound,but we can easily go out ~to space~ of the `u64` range.

Reply:
With the current ADA -> voting power conversion, I think we can actually fit around 6e8 voters each with all ADA available (~3e10) in a u64 (~1.8e19)

Reply:
What is more concerning is the actual maximum stake in the dataset (we are making an almost realistic benchmark, aren't we ;) ).

Reply:
Yes, I meant that we could allow the upper bound to be the maximum ADA available in circulation an still fit in a u64. This is still realistic in the sense that it would be impossible to have more than that, but is obviously quite strange to have that much concentrated in a single voter